### PR TITLE
executor: support hash join spill

### LINKED ISSUE #55153: Support hash join spill
## Feature Request

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### What problem does this PR solve?

Issue Number: close #55153

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Support spill of hash join v2
```

### CODE REVIEW

File: pkg/executor/join/base_join_probe.go [hunk 1]
```diff
@@ -238,15 +262,36 @@ func (j *baseJoinProbe) SetChunkForProbe(chk *chunk.Chunk) (err error) {
...
+				}
+				j.spillTmpChk[partIndex].Reset()
+			}
+
+			j.matchedRowsHeaders[logicalRowIndex] = 0
+		} else {
+			j.hashValues[partIndex] = append(j.hashValues[partIndex], posAndHashValue{hashValue: hashValue, pos: logicalRowIndex})
+		}
 	}
+
```
Comment (windtalker) [inline pkg/executor/join/base_join_probe.go:294]: Add some short path since if all the partition is spilled, no actual probe is needed?

File: pkg/executor/join/base_join_probe.go [hunk 2]
```diff
@@ -256,6 +301,134 @@ func (j *baseJoinProbe) SetChunkForProbe(chk *chunk.Chunk) (err error) {
...
+		j.usedColIdx = make([]int, 0, colNum-2)
+		for i := 0; i < colNum-2; i++ {
+			j.usedColIdx = append(j.usedColIdx, i+2)
+		}
+	}
+	j.currentChunk = chk.Prune(j.usedColIdx)
+	logicalRows := chk.NumRows()
+	j.chunkRows = logicalRows
+
+	j.usedRows = chk.Sel()
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/base_join_probe.go:324]: > I think for restored chunk, we can always makes sure that `chk.Sel()` is nil, so we don't need add extra code to handle `chk.Sel()`

okk, I will delete it.

File: pkg/executor/join/base_join_probe.go [hunk 3]
```diff
@@ -533,6 +706,13 @@ func NewJoinProbe(ctx *HashJoinCtxV2, workID uint, joinType logicalop.JoinType,
 		rUsedInOtherCondition: ctx.RUsedInOtherCondition,
 		rightAsBuildSide:      rightAsBuildSide,
 	}
+
+	probeSpillChkFieldTypes := make([]*types.FieldType, 0, len(probeChkFieldTypes)+2)
+	probeSpillChkFieldTypes = append(probeSpillChkFieldTypes, types.NewFieldType(mysql.TypeLonglong))
+	probeSpillChkFieldTypes = append(probeSpillChkFieldTypes, types.NewFieldType(mysql.TypeBit))
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/base_join_probe.go:712]: > why the type is bit?

We store serialize keys in this column, the type of serialize keys is byte so column type is set to bit.

File: pkg/executor/join/hash_join_spill.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+func (*hashJoinSpillAction) GetPriority() int64 {
+	return memory.DefSpillPriority
+}
+
+func (h *hashJoinSpillAction) Action(t *memory.Tracker) {
+	if h.actionImpl(t) {
+		return
+	}
+
+	if t.CheckExceed() && !hasEnoughDataToSpill(h.spillHelper.hashJoinExec.memTracker, t) {
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_spill.go]: > why need to check these before trigger fallback action?

Before spill flag is set to `inSpilling`, we will directly exit `actionImpl` function though spill has been triggered. In this case, triggering other executor's spill is unnecessary.

File: pkg/executor/join/hash_join_spill_helper.go [hunk 5]
```diff
@@ -0,0 +1,607 @@
...
+	helper := &hashJoinSpillHelper{hashJoinExec: hashJoinExec}
+	helper.cond = sync.NewCond(new(sync.Mutex))
+	helper.buildSpillChkFieldTypes = make([]*types.FieldType, 0, 3)
+	helper.buildSpillChkFieldTypes = append(helper.buildSpillChkFieldTypes, types.NewFieldType(mysql.TypeLonglong)) // hash value
+	helper.buildSpillChkFieldTypes = append(helper.buildSpillChkFieldTypes, types.NewFieldType(mysql.TypeBit))      // valid join key
+	helper.buildSpillChkFieldTypes = append(helper.buildSpillChkFieldTypes, types.NewFieldType(mysql.TypeBit))      // row data
+	helper.probeFieldTypes = make([]*types.FieldType, 0, 3)
+	helper.probeFieldTypes = append(helper.probeFieldTypes, types.NewFieldType(mysql.TypeLonglong)) // hash value
+	helper.probeFieldTypes = append(helper.probeFieldTypes, types.NewFieldType(mysql.TypeBit))      // serialized key
+	helper.probeFieldTypes = append(helper.probeFieldTypes, probeFieldTypes...)                     // row data
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_spill_helper.go:100]: > if `spillHelper` contains `probeFieldTypes`, why need to record `probeFieldTypes` in `JoinProbe`?

I think I should write a new function to unify the initialization of `probeFieldTypes`.

[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_spill_helper.go:100]: > if `spillHelper` contains `probeFieldTypes`, why need to record `probeFieldTypes` in `JoinProbe`?

I think I should write a new function to unify the initialization of `probeFieldTypes`.

File: pkg/executor/join/hash_join_v2.go [hunk 6]
```diff
@@ -18,32 +18,46 @@ import (
...
 	// enableHashJoinV2 is a variable used only in test
 	enableHashJoinV2 = atomic.Bool{}
 )
 
+func getHashTableLength(table *rowTable) uint64 {
+	return max(nextPowerOfTwo(table.validKeyCount()), uint64(minimalHashTableLen))
+}
+
+func getHashTableMemoryUsage(hashTableLength uint64) int64 {
+	return int64(hashTableLength) * serialization.UnsafePointerLen
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_v2.go:58]: > And why define `getHashTableLength` and `getHashTableMemoryUsage` in this file instead of `hash_table_v2.go`?

I will move them to `hash_table_v2.go`

File: pkg/executor/join/hash_join_v2.go [hunk 7]
```diff
@@ -724,42 +1222,68 @@ func (e *HashJoinV2Exec) fetchAndBuildHashTable(ctx context.Context) {
...
+	var waitForController chan struct{}
+	if e.inRestore {
+		waitForController = make(chan struct{})
+	}
+
 	wg.Add(1)
 	e.workerWg.RunWithRecover(
 		func() {
 			defer trace.StartRegion(ctx, "HashJoinBuildSideFetcher").End()
 			fetcher := e.BuildWorkers[0]
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_v2.go:1266]: > This line can be moved to L1285

done

File: pkg/executor/join/hash_join_spill_helper.go [hunk 8]
```diff
@@ -0,0 +1,607 @@
...
+
+	if h.buildRowsInDisk == nil {
+		return nil
+	}
+
+	partNum := h.hashJoinExec.partitionNumber
+	concurrency := int(h.hashJoinExec.Concurrency)
+
+	for i := 0; i < int(partNum); i++ {
+		buildInDisks := make([]*chunk.DataInDiskByChunks, 0)
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_spill_helper.go:541]: > why not
> 
> ```
> if h.spilledPartitions[i] == true {
>     // construct restorePartition
> }
> ```

done

File: pkg/executor/join/hash_join_spill_helper.go [hunk 9]
```diff
@@ -0,0 +1,607 @@
...
+		h.buildRowsInDisk = make([][]*chunk.DataInDiskByChunks, h.hashJoinExec.Concurrency)
+		h.probeRowsInDisk = make([][]*chunk.DataInDiskByChunks, h.hashJoinExec.Concurrency)
+	}
+}
+
+func (h *hashJoinSpillHelper) spillRowTableImpl(partitionsNeedSpill []int, totalReleasedMemory int64) error {
+	workerNum := len(h.hashJoinExec.BuildWorkers)
+	errChannel := make(chan error, workerNum)
+
+	waiter := &sync.WaitGroup{}
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_spill_helper.go:416]: > util.WaitGroupWrapper

`util.WaitGroupWrapper` adds counter one by one, this will cause unexpected behaviour. We need to increase counter only once.

File: pkg/executor/join/hash_join_spill_helper.go [hunk 10]
```diff
@@ -0,0 +1,607 @@
...
+	for i := 0; i < rowLen; i++ {
+		validJoinKeys[i] = byte(0)
+	}
+	for _, pos := range seg.validJoinKeyPos {
+		validJoinKeys[pos] = byte(1)
+	}
+	return validJoinKeys
+}
+
+func (h *hashJoinSpillHelper) spillSegmentsToDisk(workerID int, partID int, segments []*rowTableSegment) error {
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_spill_helper.go:335]: > `spillBuildSegmentToDisk`

done

File: pkg/executor/join/hash_join_spill_helper.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_spill_helper.go:416]: > util.WaitGroupWrapper

done

File: pkg/executor/join/row_table_builder.go [hunk 11]
```diff
@@ -148,10 +156,86 @@ func (b *rowTableBuilder) ResetBuffer(chk *chunk.Chunk) {
...
+			hasValidJoinKey = 0
+		}
+
+		var seg *rowTableSegment
+		if hasValidJoinKey != 0 {
+			newHashValue, partID, err = b.regenerateHashValueAndPartIndex(uint64(oldHashValue), hashJoinCtx.partitionMaskOffset)
+			if err != nil {
+				return err
+			}
+			seg = hashJoinCtx.hashTableContext.getCurrentRowSegment(workerID, partID, true, b.firstSegRowSizeHint)
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/row_table_builder.go:201]: > maxRowTableSegmentSize

done

File: pkg/executor/join/base_join_probe.go (hunk 1 above)
Comment (xzhangxian1008) [reply pkg/executor/join/base_join_probe.go:294]: > Add some short path since if all the partition is spilled, no actual probe is needed?

done

File: pkg/executor/join/hash_join_v2.go [hunk 12]
```diff
@@ -602,20 +1039,65 @@ func (w *ProbeWorkerV2) getNewJoinResult() (bool, *hashjoinWorkerResult) {
...
+	}()
+
+	lastRound := 0
+	for {
+		e.fetchAndBuildHashTable(ctx)
+		e.fetchAndProbeHashTable(ctx)
+
+		e.waiterWg.Wait()
+
+		e.spillHelper.spillRoundForTest = int(math.Max(float64(e.spillHelper.spillRoundForTest), float64(lastRound)))
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_v2.go:1055]: done

File: pkg/executor/join/base_join_probe.go [hunk 13]
```diff
@@ -229,6 +242,17 @@ func (j *baseJoinProbe) SetChunkForProbe(chk *chunk.Chunk) (err error) {
 			return err
 		}
 	}
+
+	// Not all sqls need spill, so we initialize it at runtime, or there will be too many unnecessary memory allocations
+	// spillTriggered can only be set in build stage, so it's ok to get it without lock
+	if j.ctx.spillHelper.isSpillTriggeredNoLock() && len(j.spillTmpChk) != int(j.ctx.partitionNumber) {
```
[the remark this answers is not in the document]
Comment (windtalker) [reply pkg/executor/join/base_join_probe.go:243]: then why not use 
```
if j.ctx.spillHelper.isSpillTriggeredNoLock() && len(j.spillTmpChk) == 0 
```

[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/base_join_probe.go]: > then why not use
> 
> ```
> if j.ctx.spillHelper.isSpillTriggeredNoLock() && len(j.spillTmpChk) == 0 
> ```

okk

File: pkg/executor/join/hash_join_v2.go [hunk 14]
```diff
@@ -602,20 +1032,65 @@ func (w *ProbeWorkerV2) getNewJoinResult() (bool, *hashjoinWorkerResult) {
...
+			return
+		}
+
+		restoredPartition := e.spillHelper.stack.pop()
+		if restoredPartition == nil {
+			// No more data to restore
+			return
+		}
+
+		if e.memTracker.BytesConsumed() != 0 {
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_v2.go:1067]: > I think `BytesConsumed` should also be set to 0 if it is not 0 here?

If `BytesConsumed` is not 0, it means that there are bugs of managing memory usage in the codes.

File: pkg/executor/join/hash_join_v2.go [hunk 15]
```diff
@@ -768,19 +1283,86 @@ func (e *HashJoinV2Exec) fetchBuildSideRows(ctx context.Context, wg *sync.WaitGr
 			wg.Done()
 		},
 	)
-	return srcChkCh
+	return srcChkCh, waitForController
+}
+
+func (e *HashJoinV2Exec) getRestoredChunkNum() int {
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_v2.go:1296]: > The function name should be `getRestoredBuildChunkNum()`?

done

File: pkg/executor/join/base_join_probe.go [hunk 16]
```diff
@@ -256,13 +298,136 @@ func (j *baseJoinProbe) SetChunkForProbe(chk *chunk.Chunk) (err error) {
...
+	if cap(j.selRows) >= logicalRows {
+		j.selRows = j.selRows[:logicalRows]
+	} else {
+		j.selRows = make([]int, 0, logicalRows)
+		for i := 0; i < logicalRows; i++ {
+			j.selRows = append(j.selRows, i)
+		}
+	}
+
+	j.usedRows = j.selRows
```
Comment (windtalker) [inline pkg/executor/join/base_join_probe.go:324]: Add check to make sure that `chk.Sel()` is nil?

File: pkg/executor/join/base_join_probe.go [hunk 17]
```diff
@@ -256,13 +302,127 @@ func (j *baseJoinProbe) SetChunkForProbe(chk *chunk.Chunk) (err error) {
...
 	if j.matchedRowsForCurrentProbeRow > 0 {
 		j.offsetAndLengthArray = append(j.offsetAndLengthArray, offsetAndLength{offset: j.usedRows[j.currentProbeRow], length: j.matchedRowsForCurrentProbeRow})
 	}
 	j.matchedRowsForCurrentProbeRow = 0
 }
 
+func (j *baseJoinProbe) ResetProbe() {
+	// We must reset this variable or gc will raise error.
+	// However, we can't explain it so far.
+	j.cachedBuildRows = make([]matchedRowInfo, batchBuildRowSize)
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/base_join_probe.go:423]: > I think you also need to reset `rowIndexInfos` if there is other conditon.

done

File: pkg/executor/join/hash_join_v2.go [hunk 18]
```diff
@@ -602,20 +1035,90 @@ func (w *ProbeWorkerV2) getNewJoinResult() (bool, *hashjoinWorkerResult) {
 	return ok, joinResult
 }
 
+func (e *HashJoinV2Exec) reset() {
+	e.spillHelper.setCanSpillFlag(true)
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_v2.go:1043]: > I think you should call `resetHashTableContextForRestore` here?

done

File: pkg/executor/join/hash_join_v2.go [hunk 19]
```diff
@@ -342,13 +683,20 @@ func (e *HashJoinV2Exec) Open(ctx context.Context) error {
 
 	e.diskTracker = disk.NewTracker(e.ID(), -1)
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_v2.go:684]: > Not introduced by this pr, but I think we need to check if `e.diskTracker` is nil like `memTracker`

done

File: tests/integrationtest/r/executor/jointest/join.result [hunk 20]
```diff
@@ -932,32 +932,6 @@ c1	c2	c1	c2
 1	2	1	2
 drop table if exists t;
 drop table if exists t1;
-create table t(c1 bigint unsigned);
```
Comment (windtalker) [inline tests/integrationtest/r/executor/jointest/join.result:935]: why remove these tests

File: pkg/executor/join/base_join_probe.go [hunk 21]
```diff
@@ -155,11 +174,6 @@ func (j *baseJoinProbe) finishCurrentLookupLoop(joinedChk *chunk.Chunk) {
 }
 
 func (j *baseJoinProbe) SetChunkForProbe(chk *chunk.Chunk) (err error) {
-	if j.currentChunk != nil {
```
Comment (windtalker) [inline pkg/executor/join/base_join_probe.go:158]: why remove this check

File: pkg/executor/join/hash_join_v2.go [hunk 22]
```diff
@@ -602,20 +1064,91 @@ func (w *ProbeWorkerV2) getNewJoinResult() (bool, *hashjoinWorkerResult) {
 	return ok, joinResult
 }
 
+func (e *HashJoinV2Exec) reset() {
+	e.spillHelper.setCanSpillFlag(true)
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_v2.go:1069]: > I think `e.spillHelper.setCanSpillFlag(true)` should be called after `e.resetHashTableContextForRestore()`?

done

File: pkg/executor/join/base_join_probe.go [hunk 23]
```diff
@@ -229,6 +243,17 @@ func (j *baseJoinProbe) SetChunkForProbe(chk *chunk.Chunk) (err error) {
...
+	if j.ctx.spillHelper.isSpillTriggered() && len(j.spillTmpChk) == 0 {
+		for i := 0; i < int(j.ctx.partitionNumber); i++ {
+			j.spillTmpChk = append(j.spillTmpChk, chunk.NewChunkWithCapacity(j.probeSpillChkFieldTypes, spillChunkSize))
+		}
+	}
+
+	j.spilledIdx = j.spilledIdx[:0]
+
 	// generate hash value
 	hash := fnv.New64()
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/base_join_probe.go:258]: > There is already a `hash` in `baseJoinProbe`, why not use a new one?

I will delete this old one.

**Review by windtalker — APPROVED** [review]
lgtm

File: pkg/executor/builder.go [hunk 24]
```diff
@@ -1658,7 +1658,7 @@ func (b *executorBuilder) buildHashJoinV2(v *plannercore.PhysicalHashJoin) exec.
 }
 
 func (b *executorBuilder) buildHashJoin(v *plannercore.PhysicalHashJoin) exec.Executor {
-	if b.ctx.GetSessionVars().UseHashJoinV2 && join.IsHashJoinV2Supported() && v.CanUseHashJoinV2() {
+	if join.IsHashJoinV2Supported() && v.CanUseHashJoinV2() {
```
Comment (windtalker) [inline pkg/executor/builder.go:1661]: need revert this change before merge

**Comment by xzhangxian1008** [comment]
Run endless tests again before merge.

File: pkg/executor/join/hash_join_spill.go [hunk 25]
```diff
@@ -0,0 +1,89 @@
...
+		h.spillHelper.bytesConsumed.Store(t.BytesConsumed())
+		h.spillHelper.bytesLimit.Store(t.GetBytesLimit())
+		return true
+	}
+
+	return false
+}
+
+func hasEnoughDataToSpill(hashJoinTracker *memory.Tracker, passedInTracker *memory.Tracker) bool {
+	return hashJoinTracker.BytesConsumed() >= passedInTracker.GetBytesLimit()/20
```
Comment (XuHuaiyu) [inline pkg/executor/join/hash_join_spill.go:88]: What does this mean?

File: pkg/executor/join/hash_join_v2.go [hunk 26]
```diff
@@ -269,7 +666,30 @@ type HashJoinV2Exec struct {
...
+	e.maxSpillRound = 1
+	totalPartitionsNum := e.partitionNumber
+	for {
+		if totalPartitionsNum > 1024 {
+			break
+		}
+		totalPartitionsNum *= e.partitionNumber
+		e.maxSpillRound++
+	}
+	e.maxSpillRound = max(1, e.maxSpillRound-1)
```
Comment (XuHuaiyu) [inline pkg/executor/join/hash_join_v2.go:692]:
```suggestion
func (e *HashJoinV2Exec) initMaxSpillRound() {
	if e.partitionNumber > 1024 {
		e.maxSpillRound = 1
		return
	}
	// Calculate the minimum number of rounds required for the total partitions to exceed 1024

	e.maxSpillRound = int(math.Ceil(math.Log(1024) / math.Log(float64(e.partitionNumber))))

```

File: pkg/executor/join/hash_join_v2.go [hunk 27]
```diff
@@ -419,29 +826,50 @@ func (e *HashJoinV2Exec) initializeForProbe() {
...
+				func() bool { return e.spillHelper.isSpillTriggered() },
+				e.ProbeSideTupleFetcher.canSkipProbeIfHashTableIsEmpty,
+				e.ProbeSideTupleFetcher.needScanRowTableAfterProbeDone,
+				e.ProbeSideTupleFetcher.shouldLimitProbeFetchSize(),
+				&e.ProbeSideTupleFetcher.hashJoinCtxBase)
+		}
+		e.workerWg.RunWithRecover(fetchProbeSideChunksFunc, e.ProbeSideTupleFetcher.handleProbeSideFetcherPanic)
+	}
+}
+
```
Comment (XuHuaiyu) [inline pkg/executor/join/hash_join_v2.go:846]: Would this be more clear?

```suggestion
func (e *HashJoinV2Exec) fetchAndProbeHashTable(ctx context.Context) {
	if e.inRestore {
		if err := <-e.buildFinished; err != nil {
			return
		}
		e.startRestoreProbeJoinWorkers(ctx)
	} else {
		e.startProbeFetcher(ctx)
		e.startNormalProbeJoinWorkers(ctx)
	}
	e.waiterWg.RunWithRecover(e.waitJoinWorkers, nil)
}

func (e *HashJoinV2Exec) startProbeFetcher(ctx context.Context) {
	fetchProbeSideChunksFunc := func() {
		defer trace.StartRegion(ctx, "HashJoinProbeSideFetcher").End()
		e.ProbeSideTupleFetcher.fetchProbeSideChunks(
			ctx,
			e.MaxChunkSize(),
			e.ProbeSideTupleFetcher.hashTableContext.hashTable.isHashTableEmpty,
			e.spillHelper.isSpillTriggered,
			e.ProbeSideTupleFetcher.canSkipProbeIfHashTableIsEmpty,
			e.ProbeSideTupleFetcher.needScanRowTableAfterProbeDone,
			e.ProbeSideTupleFetcher.shouldLimitProbeFetchSize(),
			&e.ProbeSideTupleFetcher.hashJoinCtxBase,
		)
	}
	e.workerWg.RunWithRecover(fetchProbeSideChunksFunc, e.ProbeSideTupleFetcher.handleProbeSideFetcherPanic)
}

func (e *HashJoinV2Exec) startNormalProbeJoinWorkers(ctx context.Context) {
	for i := uint(0); i < e.Concurrency; i++ {
		workerID := i
		e.workerWg.RunWithRecover(func() {
			defer trace.StartRegion(ctx, "HashJoinWorker").End()
			e.ProbeWorkers[workerID].runJoinWorker()
		}, e.ProbeWorkers[workerID].handleProbeWorkerPanic)
	}
}

func (e *HashJoinV2Exec) startRestoreProbeJoinWorkers(ctx context.Context) {
	for i := uint(0); i < e.Concurrency; i++ {
		workerID := i
		e.workerWg.RunWithRecover(func() {
			defer trace.StartRegion(ctx, "HashJoinWorker").End()
			e.ProbeWorkers[workerID].restoreAndProbe(e.restoredProbeInDisk[workerID])
		}, e.ProbeWorkers[workerID].handleProbeWorkerPanic)
	}
}

```

File: pkg/executor/join/hash_join_spill.go (hunk 25 above)
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_spill.go:88]: > What does this mean?

Like aggregation executor, hash join spill can be triggered only when it's memory usage exceeds a threoshold.

File: pkg/executor/join/hash_join_v2.go (hunk 27 above)
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_v2.go:846]: I think these two ways of writing are same.

**Comment by xzhangxian1008** [comment]
> *** Run endless tests again before merged. ***

Uncomment `TestIssue20270` before merge

File: pkg/executor/join/hash_join_v2.go [hunk 28]
```diff
@@ -228,13 +377,261 @@ type ProbeWorkerV2 struct {
...
+	for i := 0; i < chunkNum; i++ {
+		select {
+		case <-w.HashJoinCtx.closeCh:
+			return
+		default:
+		}
+		failpoint.Inject("ConsumeRandomPanic", nil)
+
+		// TODO reuse chunk
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_v2.go:411]: > This optimization is suggested to be implemented before GA. Too many small objects bring memory pressure, which might cause OOM during spilling.

okk, I will create another pr to finish this.

File: pkg/executor/join/hash_join_base.go [hunk 29]
```diff
@@ -218,10 +229,61 @@ type buildWorkerBase struct {
...
+}
+
+func syncerDone(syncer *sync.WaitGroup) {
+	if syncer != nil {
+		syncer.Done()
+	}
+}
+
+func checkAndSpillRowTableIfNeeded(fetcherAndWorkerSyncer *sync.WaitGroup, spillHelper *hashJoinSpillHelper) error {
+	if fetcherAndWorkerSyncer == nil {
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_base.go:245]: > when will this if check be true?

When it's called by hash join v1

File: pkg/executor/join/hash_join_spill_helper.go [hunk 30]
```diff
@@ -0,0 +1,585 @@
...
+
+	wg := &util.WaitGroupWrapper{}
+
+	h.setPartitionSpilled(partitionsNeedSpill)
+
+	if len(partitionsNeedSpill) == int(h.hashJoinExec.partitionNumber) {
+		h.allPartitionsSpilledForTest = true
+	}
+
+	h.spillTriggeredForTest = true
```
Comment (XuHuaiyu) [inline pkg/executor/join/hash_join_spill_helper.go:390]: Will this be better

```suggestion
if intest.InTest {
	if len(partitionsNeedSpill) == int(h.hashJoinExec.partitionNumber) {
		h.allPartitionsSpilledForTest = true
	}

	h.spillTriggeredForTest = true
}
```

**Review by XuHuaiyu — APPROVED** [review]
